nc/source/physics/collision/Manifold.cpp

```diff
@@ -0,0 +1,161 @@
+#include "Manifold.h"
+#include "physics/PhysicsConstants.h"
+
+#include <array>
+
+namespace nc::physics
+{
+    bool operator == (const Contact& a, const Contact& b)
+    {
+        return (a.normal == b.normal) &&
+               (
+                   (a.worldPointA == b.worldPointA && a.worldPointB == b.worldPointB) ||
+                   (a.worldPointA == b.worldPointB && a.worldPointB == b.worldPointA)
+               );
+    }
+
+    int ClosestAxis(float x, float y, float z, float w)
+    {
+        x = fabs(x); y = fabs(y); z = fabs(z); w = fabs(w);
+        int maxIndex = 0;
+        float maxVal = x;
+
+        if(y > maxVal)
+        {
+            maxIndex = 1;
+            maxVal = y;
+        }
+
+        if(z > maxVal)
+        {
+            maxIndex = 2;
+            maxVal = z;
+        }
+
+        if(w > maxVal)
+        {
+            maxIndex = 3;
+        }
+
+        return maxIndex;
+    }
+
+    float CalcArea4Points(const Vector3& p0, const Vector3& p1, const Vector3& p2, const Vector3& p3)
+    {
+        float mag0 = SquareMagnitude(CrossProduct(p0-p1, p2-p3));
+        float mag1 = SquareMagnitude(CrossProduct(p0-p2, p1-p3));
+        float mag2 = SquareMagnitude(CrossProduct(p0-p3, p1-p2));
+        return math::Max(math::Max(mag0, mag1), mag2);
+    }
+
+    int Manifold::AddContact(const Contact& contact)
+    {
+        int insertIndex = contacts.size();
+
+        if(insertIndex >= 4)
+        {
+            insertIndex = SortPoints(contact);
+            
+            if(insertIndex < 0)
+                insertIndex = 0;
+
+            contacts[insertIndex] = contact;
+        }
+        else
+        {
+            contacts.push_back(contact);
+        }
+
+        return insertIndex;
+    }
+
+    const Contact& Manifold::GetDeepestContact() const
+    {
+        size_t maxPenetrationIndex = 0;
+        float maxPenetration = contacts[0].depth;
```

Comment:
Ever a chance contacts is empty here?

Reply:
I don't think so, but I'll assert anyway.